ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3CompleteMultipartUploadHandler.java

```diff
@@ -0,0 +1,366 @@
+          // Do ttl update with retryExecutor. Use the blob ID returned from the router instead of the converted ID
+          // since the converted ID may be changed by the ID converter.
+          String serviceId = blobInfo.getBlobProperties().getServiceId();
+          retryExecutor.runWithRetries(retryPolicy,
+              callback -> router.updateBlobTtl(blobId, serviceId, Utils.Infinite_Time, callback,
+                  QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, false)), this::isRetriable,
+              routerTtlUpdateCallback(blobInfo, blobId));
+        } else {
+          securityService.processResponse(restRequest, restResponseChannel, blobInfo,
+              securityProcessResponseCallback());
+        }
+      }, uri, LOGGER, finalCallback);
+    }
+
+    /**
+     * After TTL update finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @param blobInfo the {@link BlobInfo} to use for security checks.
+     * @param blobId the {@link String} to use for blob id.
+     * @return a {@link Callback} to be used with {@link Router#updateBlobTtl(String, String, long)}.
+     */
+    private Callback<Void> routerTtlUpdateCallback(BlobInfo blobInfo, String blobId) {
+      return buildCallback(frontendMetrics.updateBlobTtlRouterMetrics, convertedBlobId -> {
+        // Set the named blob state to be 'READY' after the Ttl update succeed
+        String blobIdClean = RestUtils.stripSlashAndExtensionFromId(blobId);
+        NamedBlobPath namedBlobPath = NamedBlobPath.parse(RestUtils.getRequestPath(restRequest), restRequest.getArgs());
+        NamedBlobRecord record = new NamedBlobRecord(namedBlobPath.getAccountName(), namedBlobPath.getContainerName(),
+            namedBlobPath.getBlobName(), blobIdClean, Utils.Infinite_Time);
+        namedBlobDb.updateBlobTtlAndStateToReady(record).get();
+        securityService.processResponse(restRequest, restResponseChannel, blobInfo, securityProcessResponseCallback());
+      }, uri, LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link SecurityService#processResponse}, call {@code finalCallback}.
+     * @return a {@link Callback} to be used with {@link SecurityService#processResponse}.
+     */
+    private Callback<Void> securityProcessResponseCallback() {
+      return buildCallback(frontendMetrics.putBlobSecurityProcessResponseMetrics, securityCheckResult -> {
+        // Create xml response to send to user
+        RequestPath requestPath = (RequestPath) restRequest.getArgs().get(REQUEST_PATH);
+        NamedBlobPath namedBlobPath = NamedBlobPath.parse(requestPath, restRequest.getArgs());
+        String bucket = namedBlobPath.getAccountName();
+        String key = namedBlobPath.getBlobName();
+        String eTag = (String) restResponseChannel.getHeader(RestUtils.Headers.LOCATION);
```

Comment:
Location and Etag are different, right?
```
<Location>http://Example-Bucket.s3.<Region>.amazonaws.com/Example-Object</Location>
<ETag>"3858f62230ac3c915f300c664312c11f-9"</ETag>
```

And we are getting the LOCATION from the responseChannel. Do we set it before?